### PR TITLE
Verifiy authenticity of the majority of downloads.

### PR DESCRIPTION
**Pull request by RyanHir**
Use GPG and SHA256 for verification, unless said verification resources are unavailable.

In the case where National Instruments does not offer a way to verify the signature, the SHA256 is compared with the `Packages` file they provide. While the `Packages` file has a signature, it is unusable as I was unable to find the public key online so there is no verification of the Packages file, so it will become reliant on HTTPS.

The only two files I could not verify were cloog and isl, as they don't provide signature nor is HTTPS secured.

While this is not critical for usage, its a nice to have and would add on to #19 so it can have a near complete trusted source of code.

Here is the new output of download.sh, I am aware of the warnings and they can be ignored as they are from the keyserver which is also used by Ubuntu to build their packages.
```sh
ryan@think:~/devel/roborio-toolchain$ ./download.sh 
gpg: key 6EEB81F8981C74C7: "Richard Guenther <richard.guenther@gmail.com>" not changed
gpg: Total number processed: 1
gpg:              unchanged: 1
gpg: key C3126D3B4AE55E93: "Tristan Gingold <gingold@adacore.com>" not changed
gpg: Total number processed: 1
gpg:              unchanged: 1
gpg: key 980C197698C3739D: "Vincent Lefevre <vincent@vinc17.net>" not changed
gpg: Total number processed: 1
gpg:              unchanged: 1
gpg: key F7D5C9BF765C61E3: "Andreas Enge <andreas.enge@inria.fr>" not changed
gpg: Total number processed: 1
gpg:              unchanged: 1
gpg: key F3599FF828C67298: "Niels Möller <nisse@lysator.liu.se>" not changed
gpg: Total number processed: 1
gpg:              unchanged: 1
gpg: key 92EDB04BFF325CF3: "Joel Brobecker <brobecker@adacore.com>" not changed
gpg: Total number processed: 1
gpg:              unchanged: 1
gpg: key B00BC66A401A1600: "Sebastian Pipping (sping) <sping@gentoo.org>" not changed
gpg: Total number processed: 1
gpg:              unchanged: 1
2020-10-20 15:01:35 URL:https://ftp.gnu.org/gnu/gcc/gcc-7.3.0/gcc-7.3.0.tar.gz [112201917/112201917] -> "gcc-7.3.0.tar.gz" [1]
2020-10-20 15:01:35 URL:https://ftp.gnu.org/gnu/gcc/gcc-7.3.0/gcc-7.3.0.tar.gz.sig [280/280] -> "gcc-7.3.0.tar.gz.sig" [1]
gpg: assuming signed data in 'gcc-7.3.0.tar.gz'
gpg: Signature made Thu 25 Jan 2018 03:21:04 AM CST
gpg:                using RSA key 3AB00996FC26A641
gpg: Good signature from "Richard Guenther <richard.guenther@gmail.com>" [unknown]
gpg:                 aka "Richard Guenther (Work) <rguenther@suse.de>" [unknown]
gpg:                 aka "Richard Guenther <richard.guenther@gmx.net>" [unknown]
gpg:                 aka "Richard Guenther (GCC) <rguenth@gcc.gnu.org>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 1397 5A70 E63C 361C 73AE  69EF 6EEB 81F8 981C 74C7
     Subkey fingerprint: 7F74 F97C 1034 68EE 5D75  0B58 3AB0 0996 FC26 A641
2020-10-20 15:01:41 URL:https://ftp.gnu.org/gnu/binutils/binutils-2.28.tar.bz2 [26556365/26556365] -> "binutils-2.28.tar.bz2" [1]
2020-10-20 15:01:41 URL:https://ftp.gnu.org/gnu/binutils/binutils-2.28.tar.bz2.sig [194/194] -> "binutils-2.28.tar.bz2.sig" [1]
gpg: assuming signed data in 'binutils-2.28.tar.bz2'
gpg: Signature made Thu 02 Mar 2017 03:21:06 AM CST
gpg:                using DSA key C3126D3B4AE55E93
gpg: Good signature from "Tristan Gingold <gingold@adacore.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: EAF1 C276 A747 E9ED 8621  0CBA C312 6D3B 4AE5 5E93
2020-10-20 15:01:42 URL:https://ftp.gnu.org/gnu/mpfr/mpfr-3.1.5.tar.bz2 [1279489/1279489] -> "mpfr-3.1.5.tar.bz2" [1]
2020-10-20 15:01:43 URL:https://ftp.gnu.org/gnu/mpfr/mpfr-3.1.5.tar.bz2.sig [72/72] -> "mpfr-3.1.5.tar.bz2.sig" [1]
gpg: assuming signed data in 'mpfr-3.1.5.tar.bz2'
gpg: Signature made Tue 27 Sep 2016 06:09:34 AM CDT
gpg:                using DSA key 980C197698C3739D
gpg: Good signature from "Vincent Lefevre <vincent@vinc17.net>" [unknown]
gpg:                 aka "Vincent Lefevre <vincent@vinc17.org>" [unknown]
gpg:                 aka "Vincent Lefèvre <vincent@vinc17.net>" [unknown]
gpg:                 aka "Vincent Lefevre <Vincent.Lefevre@inria.fr>" [unknown]
gpg:                 aka "Vincent Lefèvre <Vincent.Lefevre@ens-lyon.fr>" [unknown]
gpg:                 aka "Vincent Lefèvre <Vincent.Lefevre@inria.fr>" [unknown]
gpg:                 aka "Vincent Lefèvre <vincent@vinc17.org>" [unknown]
gpg:                 aka "Vincent Lefevre <Vincent.Lefevre@ens-lyon.fr>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 07F3 DBBE CC1A 3960 5078  094D 980C 1976 98C3 739D
2020-10-20 15:01:43 URL:https://ftp.gnu.org/gnu/mpc/mpc-1.0.3.tar.gz [669925/669925] -> "mpc-1.0.3.tar.gz" [1]
2020-10-20 15:01:43 URL:https://ftp.gnu.org/gnu/mpc/mpc-1.0.3.tar.gz.sig [213/213] -> "mpc-1.0.3.tar.gz.sig" [1]
gpg: assuming signed data in 'mpc-1.0.3.tar.gz'
gpg: Signature made Mon 16 Feb 2015 10:16:34 AM CST
gpg:                using DSA key F7D5C9BF765C61E3
gpg: Good signature from "Andreas Enge <andreas.enge@inria.fr>" [unknown]
gpg:                 aka "Andreas Enge <andreas@enge.fr>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: AD17 A21E F8AE D8F1 CC02  DBD9 F7D5 C9BF 765C 61E3
2020-10-20 15:01:44 URL:https://ftp.gnu.org/gnu/gmp/gmp-6.1.2.tar.bz2 [2386766/2386766] -> "gmp-6.1.2.tar.bz2" [1]
2020-10-20 15:01:45 URL:https://ftp.gnu.org/gnu/gmp/gmp-6.1.2.tar.bz2.sig [351/351] -> "gmp-6.1.2.tar.bz2.sig" [1]
gpg: assuming signed data in 'gmp-6.1.2.tar.bz2'
gpg: Signature made Sun 18 Dec 2016 02:18:25 PM CST
gpg:                using RSA key F3599FF828C67298
gpg: Good signature from "Niels Möller <nisse@lysator.liu.se>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 343C 2FF0 FBEE 5EC2 EDBE  F399 F359 9FF8 28C6 7298
2020-10-20 15:01:47 URL:https://ftp.gnu.org/gnu/gdb/gdb-8.0.1.tar.gz [36359351/36359351] -> "gdb-8.0.1.tar.gz" [1]
2020-10-20 15:01:48 URL:https://ftp.gnu.org/gnu/gdb/gdb-8.0.1.tar.gz.sig [72/72] -> "gdb-8.0.1.tar.gz.sig" [1]
gpg: assuming signed data in 'gdb-8.0.1.tar.gz'
gpg: Signature made Thu 07 Sep 2017 10:00:09 AM CDT
gpg:                using DSA key 92EDB04BFF325CF3
gpg: Good signature from "Joel Brobecker <brobecker@adacore.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: F40A DB90 2B24 264A A42E  50BF 92ED B04B FF32 5CF3
2020-10-20 15:01:49 URL:https://cfhcable.dl.sourceforge.net/project/expat/expat/2.2.0/expat-2.2.0.tar.bz2 [414352/414352] -> "expat-2.2.0.tar.bz2" [1]
https://sourceforge.net/projects/expat/files/expat/2.2.0/expat-2.2.0.tar.bz2.asc:
2020-10-20 15:01:50 ERROR 404: Not Found.
2020-10-20 15:01:51 URL:https://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/Packages [5714821/5714821] -> "Packages" [1]
2020-10-20 15:01:52 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/gcc_7.3.0-r0.13_cortexa9-vfpv3.ipk [14788300/14788300] -> "gcc_7.3.0-r0.13_cortexa9-vfpv3.ipk" [1]
gcc_7.3.0-r0.13_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:52 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libgcc1_7.3.0-r0.16_cortexa9-vfpv3.ipk [44080/44080] -> "libgcc1_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libgcc1_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:53 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libgcc-s-dbg_7.3.0-r0.16_cortexa9-vfpv3.ipk [283188/283188] -> "libgcc-s-dbg_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libgcc-s-dbg_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:53 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libgcc-s-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk [2821680/2821680] -> "libgcc-s-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libgcc-s-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:54 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libstdc++6_7.3.0-r0.16_cortexa9-vfpv3.ipk [419092/419092] -> "libstdc++6_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libstdc++6_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:54 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libstdc++-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk [1360286/1360286] -> "libstdc++-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libstdc++-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:55 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libatomic1_7.3.0-r0.16_cortexa9-vfpv3.ipk [6376/6376] -> "libatomic1_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libatomic1_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:55 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libatomic-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk [788/788] -> "libatomic-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libatomic-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:55 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libgomp1_7.3.0-r0.16_cortexa9-vfpv3.ipk [75404/75404] -> "libgomp1_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libgomp1_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:55 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libgomp-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk [3594/3594] -> "libgomp-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libgomp-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:56 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libitm1_7.3.0-r0.16_cortexa9-vfpv3.ipk [26518/26518] -> "libitm1_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libitm1_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:56 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libitm-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk [948/948] -> "libitm-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libitm-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:56 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libssp0_7.3.0-r0.16_cortexa9-vfpv3.ipk [3836/3836] -> "libssp0_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libssp0_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:56 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libssp-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk [4474/4474] -> "libssp-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
libssp-dev_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:57 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/gcc-runtime-dbg_7.3.0-r0.16_cortexa9-vfpv3.ipk [3370838/3370838] -> "gcc-runtime-dbg_7.3.0-r0.16_cortexa9-vfpv3.ipk" [1]
gcc-runtime-dbg_7.3.0-r0.16_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:58 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libc6_2.24-r0.79_cortexa9-vfpv3.ipk [1447874/1447874] -> "libc6_2.24-r0.79_cortexa9-vfpv3.ipk" [1]
libc6_2.24-r0.79_cortexa9-vfpv3.ipk: OK
2020-10-20 15:01:59 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libc6-dbg_2.24-r0.79_cortexa9-vfpv3.ipk [14992204/14992204] -> "libc6-dbg_2.24-r0.79_cortexa9-vfpv3.ipk" [1]
libc6-dbg_2.24-r0.79_cortexa9-vfpv3.ipk: OK
2020-10-20 15:02:00 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libc6-dev_2.24-r0.79_cortexa9-vfpv3.ipk [536304/536304] -> "libc6-dev_2.24-r0.79_cortexa9-vfpv3.ipk" [1]
libc6-dev_2.24-r0.79_cortexa9-vfpv3.ipk: OK
2020-10-20 15:02:00 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/linux-libc-headers-dev_4.15.7-r0.13_cortexa9-vfpv3.ipk [1239170/1239170] -> "linux-libc-headers-dev_4.15.7-r0.13_cortexa9-vfpv3.ipk" [1]
linux-libc-headers-dev_4.15.7-r0.13_cortexa9-vfpv3.ipk: OK
2020-10-20 15:02:00 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libcidn1_2.24-r0.79_cortexa9-vfpv3.ipk [56526/56526] -> "libcidn1_2.24-r0.79_cortexa9-vfpv3.ipk" [1]
libcidn1_2.24-r0.79_cortexa9-vfpv3.ipk: OK
2020-10-20 15:02:01 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libc6-thread-db_2.24-r0.79_cortexa9-vfpv3.ipk [11414/11414] -> "libc6-thread-db_2.24-r0.79_cortexa9-vfpv3.ipk" [1]
libc6-thread-db_2.24-r0.79_cortexa9-vfpv3.ipk: OK
2020-10-20 15:02:01 URL:http://download.ni.com/ni-linux-rt/feeds/2019/arm/cortexa9-vfpv3/libc6-extra-nss_2.24-r0.79_cortexa9-vfpv3.ipk [50404/50404] -> "libc6-extra-nss_2.24-r0.79_cortexa9-vfpv3.ipk" [1]
libc6-extra-nss_2.24-r0.79_cortexa9-vfpv3.ipk: OK
2020-10-20 15:02:03 URL:http://www.bastoul.net/cloog/pages/download/cloog-0.18.4.tar.gz [4796456/4796456] -> "cloog-0.18.4.tar.gz" [1]
2020-10-20 15:02:05 URL:http://isl.gforge.inria.fr/isl-0.19.tar.bz2 [1701593/1701593] -> "isl-0.19.tar.bz2" [1]
FINISHED --2020-10-20 15:02:05--
Total wall clock time: 3.5s
Downloaded: 2 files, 6.2M in 2.7s (2.29 MB/s)
```

### CODE REVIEW

File: tools/verify.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+function signed-ni() {
+    BASE_FILE=${1/*\//}
+    PACKAGE=${BASE_FILE//_*/}
+    BLOCK=$(sed -e '/Package: '"$PACKAGE"'$/,/Priority/!d' < Packages) || { echo "Could Not Open Packages"; return; }
+    SHA=$(echo "$BLOCK" | grep 'SHA256' | sed -e 's/.*: //g')
+    wget -nc -nv $1 || { echo "Could Not Download $BASE_FILE"; exit 1; }
+
+    # Requires two spaces
+    echo "$SHA  $BASE_FILE" | shasum -a 256 -c - || { echo "Cannot Verify $BASE_FILE Download"; exit 1; }
+}
```
Comment (ThadHouse) [inline tools/verify.sh]: new line at end of file.

File: tools/verify.sh [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#! /bin/bash
+
+function signed() {
+    BASE_FILE=${2/*\//}
+    wget -nc -nv "$2" || { echo "Could Not Download $BASE_FILE"; exit 1; }
+    wget -nc -nv "$2.$1" || return 0 # Signuature missing, skip
```
Comment (ThadHouse) [inline tools/verify.sh]: spelling

File: download.sh [hunk 3]
```diff
@@ -1,35 +1,54 @@
-#!/bin/zsh
```
Comment (ThadHouse) [inline download.sh:1]: iirc, we have a specific reason to want this to be zsh, and I would prefer to not change it.

Comment (RyanHir) [reply download.sh:1]: there are some files that have zsh specific features in the project but not this file, but there are bash specific ones in this file. Zsh should be able to support it so I will let that go.

Comment (RyanHir) [reply download.sh:1]: I will wait for the CI to complete the download section to close this

Comment (RyanHir) [reply download.sh:1]: nvm, I ran it with zsh on my laptop and it worked

**Comment by RyanHir** [comment]
Notice for failing builds, the CIs are failing at artifact upload and is due to an outage at GitHub

https://github.com/actions/upload-artifact/issues/71
https://www.githubstatus.com/incidents/k46g5dn2c4lt

**Review by PeterJohnson — APPROVED** [review]

**Pull request merged**